src/nvim/mbyte.c

```diff
@@ -528,6 +528,76 @@ int utf_ptr2cells(const char *p)
   return 1;
 }
 
+/// Convert a UTF-8 byte sequence to a character number.
+/// Doesn't handle ascii! only multibyte and illegal sequences.
+///
+/// @param[in]  p      String to convert.
+/// @param[in]  len    Length of the character in bytes, 0 or 1 if illegal.
+///
+/// @return Unicode codepoint. A negative value When the sequence is illegal.
+int32_t utf_ptr2CharInfo_impl(uint8_t const *p, uintptr_t const len)
+  FUNC_ATTR_PURE FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+// uint8_t is a reminder for clang to use smaller cmp
+#define CHECK if (EXPECT((uint8_t)(cur & 0xC0U) != 0x80U, false)) goto errorret
```

Comment:
```suggestion
#define CHECK if (EXPECT((uint8_t)(cur & 0xC0U) != 0x80U, false)) return -1
```

Reply:
Also consider surrounding this with `do {` and `} while (0)`

Reply:
If `corr` is not used in all return paths, the compiler would move its read to the normal return instead of reading the value at the start of the function. This could increase latency if many different memory locations are read between character decoding. This can happen in `utf_char_compisinglike()`, and `charsize_regular()` when `linebreak` is on or there is virtual text.

I wasn't able to measure any significant difference in my benchmarks